<center> <h1> Задача 1.12 </h1> </center>

Определить оптимальный шаг $h = const$ формулы численного дифференцирования

$$f'(x) \approx \frac{\left(f(x + h) - f(x - h)\right)}{2h}, \quad \max\limits_{[x - h, x + h]}{|f'''(x)|} \leq 100,$$

если абсолютная погрешность при задании $f(x), f(x \pm h)$ не превосходит $\Delta = 0.1$

<center> <h1> Решение </h1> </center>

Применяя формулу Тейлора с остаточным членом в форме Лагранжа, получим

$$f(x + h) = f(x) + hf'(x) + \frac{h^2}{2}f''(x) + \frac{h^3}{3!}f'''(\xi)$$
$$f(x - h) = f(x) - hf'(x) + \frac{h^2}{2}f''(x) - \frac{h^3}{3!}f'''(\zeta)$$

Тогда

$$\left|f'(x) - \frac{f(x + h) - f(x - h)}{2h}\right| = \left|f'(x) - \frac{2hf'(x) + \frac{Mh^3}{3}}{2h}\right| = \frac{|M|h^2}{6},$$

где $|M| = \frac{|f'''(\xi) + f'''(\zeta)|}{2} \leq 100$.

Суммарная погрешность равна

$$\frac{h^2|M|}{6} + \frac{\Delta}{h}$$

Найдём экстремум данной функции
$$g'(h) = \frac{|M|h}{3} - \frac{\Delta}{h^2} = 0 \implies h = \sqrt[3]{\frac{3\Delta}{M}} = 0.1\sqrt[3]{3}$$

### Ответ: $h = 0.1\sqrt[3]{3}$

<center> <h1> Задача 1.17 </h1> </center>

Пусть задана последовательность чисел $x_n$,  $n = 0, 1, 2, ... $, причем $x_{n + 1} - 5x_n = 4$ , a $x_0$ известно с относительной погрешностью $10^{-6}$ .
При каких значениях $x_0$ относительная погрешность при вычислении $x_n$ будет быстро возрастать с ростом $n$? 

<center> <h1> Решение </h1> </center>
Для соотношения $x_{n + 1} - 5x_n = 4$ можно записать решение для произвольного $n$ в общем виде: $x_n = C5^n - 1$, где константа $C$ определяется из начальных данных: $C = x_0 + 1$. Теперь видно, что $x_0 = -1$ - специальный случай, когда $x_n$ не растут с ростом $n$. Небольшое отклонение приводит к быстро возрастаюшей относительной погрешности.

### Ответ: $x_0 = -1$

<center> <h1> Задача 2.1.1 </h1> </center>

Доказать, что выражение $\max\limits_{k} \left(d_k|x_k|\right), d_k > 0, k = \overline{1, n}$, есть норма вектора.

<center> <h1> Решение </h1> </center>

Проверим выполнение трёх условий:

1) $$\max\limits_{k} \left(d_k|x_k|\right) \geq 0 ; \quad \max\limits_{k} \left(d_k|x_k|\right) = 0 \iff \forall k \rightarrow x_k = 0$$
2) $$\max\limits_{k} \left(d_k|\alpha x_k|\right) = |\alpha|\max\limits_{k} \left(d_k|x_k|\right)$$
3) $$\max\limits_{k} \left(d_k|x_k + y_k|\right) \leq \max\limits_{k} \left(d_k(|x_k| + |y_k|\right) \leq \max\limits_{k} \left(d_k|x_k|\right) + \max\limits_{k} \left(d_k|y_k|\right)$$

<center> <h1> Задача 2.1.19 </h1> </center>

Найти норму матрицы $\textbf{A}$, подчинённую векторной норме $\sum_{k = 1}^{n}d_k|x_k|, d_k > 0, k = \overline{1, n}$.

<center> <h1> Решение </h1> </center>

По определению,

$$||A|| = \sup\limits_{\vec{x} \neq 0}\frac{||\textbf{A}\vec{x}||}{||\vec{x}||}$$
<!-- $$\sup\limits_{\vec{x} \neq 0} \frac{\sum_1^n d_k|(\textbf{A}\vec{x})_k|}{\sum_1^n d_k|x_k|} = \sup\limits_{\vec{x} \neq 0} \frac{\sum_{k = 1}^n d_k|\sum_{j = 1}^n a_{kj}x_j|}{\sum_1^n d_k|x_k|} = $$
$$= \sup\limits_{\vec{x} \neq 0} \frac{\sum_{k = 1}^n\sum_{j = 1}^n d_k| a_{kj}x_j|}{\sum_1^n d_k|x_k|} = \sup\limits_{\vec{x} \neq 0} \frac{\sum_{j = 1}^n d_j\sum_{k = 1}^n | a_{jk}x_k|}{\sum_1^n d_k|x_k|}$$ -->

Заметим, что $$||x|| = ||Dx||_1, \quad \text{где } D = \operatorname{diag}{(d_1, ..., d_n)}$$

Тогда получим

$$||A|| = \sup\limits_{\vec{x} \neq 0}\frac{||\textbf{A}\vec{x}||}{||\vec{x}||} = \sup\limits_{\vec{x} \neq 0} \frac{||DAx||_1}{||Dx||_1} \leq ||DAD^{-1}||_1$$

Для того, чтобы показать, что равенство достигается, достаточно взять $\vec{x} = (\frac{1}{d_1}, ..., \frac{1}{d_n})^T = D^{-1}(1, ..., 1)^T$. В этом случае имеем:

$$||A|| = \sup\frac{||DAD^{-1} (1, ..., 1)^T||_1}{||DD^{-1}(1, ..., 1)^T||_1} = ||DAD^{-1}||_1$$

<center> <h1> Задача 2.1.33 </h1> </center> 

При какой правой части $b$ системы $\textbf{A}\vec{x} = \vec{b}$, $A = 
\begin{bmatrix}
    9 & 3 \\
    -3 & -9 \\
\end{bmatrix}
$ и 
произвольной погрешности $\Delta \vec{b}$ максимальна (минимальна) величина $\nu_i$ в 
оценке $\frac{\Vert\Delta \vec{x}\Vert_i}{\Vert\vec{x}\Vert_i} \leq \nu_i\frac{\Vert\Delta\vec{b}\Vert_i}{\Vert\vec{b}\Vert_i}$ относительной ошибки решения? 

<center> <h1> Решение </h1> </center> 

По определению,

$$\nu_i = \frac{\Vert\textbf{A}^{-1}\Vert_i \Vert\vec{b}\Vert_i}{\Vert\textbf{A}^{-1}\vec{b}\Vert_i}$$

$\textbf{A}^{-1} = \frac{1}{24}
\begin{bmatrix}
    3 & 1 \\
    -1 & -3 \\
\end{bmatrix}$

$\Vert\textbf{A}^{-1}\Vert_i = \frac{1}{6}$, $i = 1, 2, 3$

$\Vert\textbf{A}^{-1}\vec{b}\Vert_1 = \max\limits_{\vec{b} \neq 0} (|3b_1 + b_2|, |b_1 + 3b_2|)$

$\Vert\textbf{A}^{-1}\vec{b}\Vert_2 = |3b_1 + b_2| + |b_1 + 3b_2|$

$\Vert\textbf{A}^{-1}\vec{b}\Vert_3 = \sqrt{|3b_1 + b_2|^2 + |b_1 + 3b_2|^2}$

WLOG предположим $|b_1| \geq |b_2|$.

Запишем $\nu_i$:

$$\nu_1 = 4\frac{\max\limits_{\vec{b} \neq 0}(|b_1|, |b_2|)}{\max\limits_{\vec{b \neq 0}} (|3b_1 + b_2|, |b_1 + 3b_2|)}$$

$$\nu_2 = 4\frac{|b_1| + |b_2|}{|3b_1 + b_2| + |b_1 + 3b_2|}$$

$$\nu_3 = 4\sqrt{\frac{b_1^2 + b_2^2}{(3b_1 + b_2)^2 + (b_1 + 3b_2)^2}}$$

1) $\nu_1$:

$$ 1   _{b_2 = b_1}\leq 4\frac{1}{\max\limits_{\vec{b \neq 0}} (|3 + \frac{b_2}{b_1}|, |1 + 3\frac{b_2}{b_1}|)} \leq_{b_2 = -b_1} 2)$$

2) $\nu_2$:

$$\nu_2 \geq 4\frac{|b_1| + |b_2|}{4|b_1| + 4|b_2|} =_{b_1 = b_2} 1$$

$$\nu_2 \leq 4\frac{1 + \frac{|b_2|}{|b_1|}}{|3 + \frac{b_2}{b_1}| + |1 + 3\frac{b_2}{b_1}|} \leq_{b_2 = -b_1} 2$$

3) $\nu_3$:

$$\nu_3 = 4\sqrt{\frac{1}{10 + 6\frac{2b_1b_2}{b_1^2 + b_2^2}}} \implies 1 _{b_2 = b_1}\leq \nu_3 \leq_{b_2 = -b_1} 2$$

<center> <h1> Задача 2.2.1 </h1> </center>

Решить СЛАУ $\textbf{A}\vec{x} = \vec{b}$ методом Гаусса с выбором главного элемента:

$\textbf{A} = 
\begin{bmatrix}
    1 & 2 & 0 \\
    3 & 4 & 5 \\
    0 & 6 & 7 \\
\end{bmatrix}
,
\vec{b} = 
\begin{bmatrix}
    5 \\
    6 \\
    5 \\
\end{bmatrix}
\quad
\textbf{A} = 
\begin{bmatrix}
    2 & 3 & 0 \\
    4 & 5 & 6 \\
    0 & 7 & 8 \\
\end{bmatrix}
,
\vec{b} = 
\begin{bmatrix}
    8 \\
    8 \\
    6 \\
\end{bmatrix}
\quad
\textbf{A} = 
\begin{bmatrix}
    3 & 4 & 0 \\
    5 & 6 & 7 \\
    0 & 8 & 9 \\
\end{bmatrix}
,
\vec{b} = 
\begin{bmatrix}
    11 \\
    10 \\
    7 \\
\end{bmatrix}$

<center> <h1> Решение </h1> </center>


In [1]:
import numpy as np
from system import System


%load_ext autoreload
%autoreload 2

In [2]:
A_1 = np.array(
    [
        [1, 2, 0],
        [3, 4, 5],
        [0, 6, 7]
    ]
)
b_1 = np.array(
    [5, 6, 5]
)

In [3]:
slae = System(A_1, b_1)
x_1 = slae.solve_gauss()
x_1

array([ 1.,  2., -1.])

In [4]:
A_2 = np.array(
    [
        [2, 3, 0],
        [4, 5, 6],
        [0, 7, 8]
    ]
)
b_2 = np.array(
    [8, 8, 6]
)

In [5]:
slae = System(A_2, b_2)
x_2 = slae.solve_gauss()
x_2

array([ 1.,  2., -1.])

In [6]:
A_3 = np.array(
    [
        [3, 4, 0],
        [5, 6, 7],
        [0, 8, 9]
    ]
)
b_3 = np.array(
    [11, 10, 7]
)

In [7]:
slae = System(A_3, b_3)
x_3 = slae.solve_gauss()
x_3

array([ 1.,  2., -1.])

<center> <h1> Задача 2.2.13 </h1> </center>

Решить СЛАУ $\textbf{A}\vec{x} = \vec{b}$ методом ортогонализации по строкам:

$\textbf{A} = 
\begin{bmatrix}
    1 & 2 \\
    3 & 4 \\
\end{bmatrix}
,
\vec{b} = 
\begin{bmatrix}
    3 \\
    5 \\
\end{bmatrix}
\quad
\textbf{A} = 
\begin{bmatrix}
    3 & 4 \\
    5 & 6 \\
\end{bmatrix}
,
\vec{b} = 
\begin{bmatrix}
    2 \\
    4 \\
\end{bmatrix}
\quad
\textbf{A} = 
\begin{bmatrix}
    5 & 6 \\
    7 & 8 \\
\end{bmatrix}
,
\vec{b} = 
\begin{bmatrix}
    9 \\
    13 \\
\end{bmatrix}$

<center> <h1> Решение </h1> </center>

In [8]:
A_1 = np.array(
    [
        [1, 2],
        [3, 4]
    ]
)
b_1 = np.array(
    [3, 5]
)

In [9]:
slae = System(A_1, b_1)
x_1 = slae.solve_ortogonalization()
print(x_1)

[-1.  2.]


In [10]:
A_2 = np.array(
    [
        [3, 4],
        [5, 6]
    ]
)
b_2 = np.array(
    [2, 4]
)

In [11]:
slae = System(A_2, b_2)
x_2 = slae.solve_ortogonalization()
print(x_2)

[ 2. -1.]


In [12]:
A_3 = np.array(
    [
        [5, 6],
        [7, 8]
    ]
)
b_3 = np.array(
    [9, 13]
)

In [13]:
slae = System(A_3, b_3)
x_3 = slae.solve_ortogonalization()
print(x_3)

[ 3. -1.]


<center> <h1> Задача 2.3.2 </h1> </center>

Произвести одну итерацию решения СЛАУ $\textbf{A}\vec{x} = \vec{b}$, $\vec{b} = (-1, 1, 1, -1)^T$, с начальным приближением $\vec{x}^{(0)} = (1, 1, 1, 1)^T$ методом Зейделя:

$\textbf{A} = 
\begin{bmatrix}
    2  & -1 &    &    \\
    -1 & 2  & -1  &    \\
       & -1 & 2  & -1 \\
       &    & -1 & 2  \\
\end{bmatrix}
\quad
\textbf{A} = 
\begin{bmatrix}
    1  & -1 &    &    \\
    -1 & 2  & -1  &    \\
       & -1 & 2  & -1 \\
       &    & -1 & 2  \\
\end{bmatrix}
\quad
\textbf{A} = 
\begin{bmatrix}
    2  & -1 &    & 1  \\
    -1 & 2  & -1  &    \\
       & -1 & 2  & -1 \\
    1  &    & -1 & 2  \\
\end{bmatrix}
$

<center> <h1> Решение </h1> </center>

In [14]:
b = np.array(
    [-1, 1, 1, -1]
)
x_0 = np.array(
    [1, 1, 1, 1]
)

In [15]:
A_1 = np.array(
    [
        [2, -1, 0, 0],
        [-1, 2, -1, 0],
        [0, -1, 2, -1],
        [0, 0, -1, 2]
    ]
)
A_2 = np.array(
    [
        [1, -1, 0, 0],
        [-1, 2, -1, 0],
        [0, -1, 2, -1],
        [0, 0, -1, 2]
    ]
)
A_3 = np.array(
    [
        [2, -1, 0, 1],
        [-1, 2, -1, 0],
        [0, -1, 2, -1],
        [1, 0, -1, 2]
    ]
)

In [16]:
slae = System(A_1, b)
slae.solve_seidel(x_0 = x_0, iter_num = 1)

array([0.  , 1.  , 1.5 , 0.25])

In [17]:
slae = System(A_2, b)
slae.solve_seidel(x_0 = x_0, iter_num = 1)

array([0.  , 1.  , 1.5 , 0.25])

In [18]:
slae = System(A_3, b)
slae.solve_seidel(x_0 = x_0, iter_num = 1)

array([-0.5   ,  0.75  ,  1.375 ,  0.4375])

<center> <h1> Задача 2.3.7 </h1> </center>

Будет ли метод Зейделя решения СЛАУ $\textbf{A}\vec{x} = \vec{b}$,

$\textbf{A} = 
\begin{bmatrix}
    10 & -1 &  2 & -3 \\
    -1 & 10 & -2 &  3 \\
     2 & -2 & 10 & -3 \\
    -3 &  3 & -3 & 10 \\
\end{bmatrix}
,
\vec{b} = 
\begin{bmatrix}
    1 \\
    2 \\
    3 \\
    4 \\
\end{bmatrix}
\quad
\textbf{A} = 
\begin{bmatrix}
    10 & -1 &  2 & -3 \\
    -1 & 10 & -2 &  3 \\
     2 & -2 & 10 & -3 \\
    -3 &  3 & -3 & 10 \\
\end{bmatrix}
,
\vec{b} = 
\begin{bmatrix}
    4 \\
    3 \\
    2 \\
    1 \\
\end{bmatrix}
$

сходящимся при начальном приближении $\vec{x}^{(0)} = (4, 3, 2, 1)^T$?

<center> <h1> Решение </h1> </center>

Заметим, что $A = A^T$. Проверим положительную определённость $A$ с помощью критерия Сильвестра:

$$\Delta_1 = 10 > 0$$


$$\Delta_2 = 
\begin{vmatrix}
    10 & -1 \\
    -1 & 10 \\
\end{vmatrix} = 99 > 0$$

$$\Delta_3 = 
\begin{vmatrix}
    10 & -1 &  2 \\
    -1 & 10 & -2 \\
     2 & -2 & 10 \\
\end{vmatrix} = 1000 + 8 - 40 - 10 - 40 > 0$$

$$\Delta_4 = 
\begin{vmatrix}
    10 & -1 &  2 & -3 \\
    -1 & 10 & -2 &  3 \\
     2 & -2 & 10 & -3 \\
    -3 &  3 & -3 & 10 \\
\end{vmatrix}$$

$\Delta_4$ считать вручную считать придётся долго (мы можем предложить метод разложения по столбцам/строкам), поэтому пусть его посчитает программа.

In [19]:
from IPython.display import display, Math

A = np.array(
    [
        [10, -1, 2, -3],
        [-1, 10, -2, 3],
        [2, -2, 10, -3],
        [-3, 3, -3, 10]
    ]
)
display(Math(f"\Delta_4 = {np.linalg.det(A):.0f}"))

<IPython.core.display.Math object>

То есть $\Delta_4 > 0$. В таком случае, из критерия Сильвестра следует, что $\textbf{A}$ - положительно определённая матрица. Тогда $\textbf{A}\vec{x} = \vec{b}$ сходится $\forall \vec{x}^{(0)} \forall \vec{b}$.

<center> <h1> Задача 2.3.13 </h1> </center>

При каких значениях $\alpha$, $\beta$ выполняется необходимое и достаточное условие сходимости метода Зейделя решения СЛАУ $\textbf{A}\vec{x} = \vec{b}$?

$\textbf{A} =
\begin{bmatrix}
    \alpha & \beta & \alpha \\
    \beta & \alpha & \beta \\
    0 & \beta & \alpha
\end{bmatrix}
\quad
\textbf{A} = 
\begin{bmatrix}
    0 & \beta & \alpha \\
    \beta & \alpha & \beta \\
    \alpha & \beta & \alpha
\end{bmatrix}
$

<center> <h1> Решение </h1> </center>

$$\forall \lambda : \det(\lambda(\textbf{L} + \textbf{D}) + \textbf{U}) = 0 \rightarrow |\lambda| < 1$$

1)
$$
\begin{vmatrix}
    \lambda\alpha & \beta & \alpha \\
    \lambda\beta & \lambda\alpha & \beta \\
    0 & \lambda\beta & \lambda\alpha
\end{vmatrix} = 0 \implies \lambda^3\alpha^3  + \lambda^2(\alpha\beta^2 - 2\alpha\beta^2) = \alpha^3\lambda^2(\lambda - \frac{\beta^2}{\alpha^2}) = 0 \implies \lambda_{1, 2} = 0, \lambda_3 = \frac{\beta^2}{\alpha^2} < 1 \implies |\beta| < |\alpha| 
$$

2)
$$
\begin{vmatrix}
    0 & \beta & \alpha \\
    \lambda\beta & \lambda\alpha & \beta \\
    \lambda\alpha & \lambda\beta & \lambda\alpha
\end{vmatrix} = 0 \implies \alpha\lambda(\beta^2 + \lambda\beta^2 - \lambda\alpha^2 - \lambda\beta^2) = 0 \implies \lambda_{1} = 0, \lambda_2 = \frac{\beta^2}{\alpha^2} < 1 \implies |\beta| < |\alpha|
$$

<center> <h1> Задача 2.3.19 </h1> </center>

Будет ли метод $\vec{x}^{(m + 1)} = (\textbf{E} - \tau\textbf{A})\vec{x}^{(m)} + \tau\vec{b}$ решения задач

$\textbf{A} = 
\begin{bmatrix}
    2 & -1 & 0 \\
    -1 & 2 & -1 \\
    0 & -1 & 2 \\
\end{bmatrix}
\quad
\textbf{A} = 
\begin{bmatrix}
    2 & -1 & -1 \\
    -1 & 2 & -1 \\
    1 & -1 & 2 \\
\end{bmatrix}
\quad
\textbf{A} = 
\begin{bmatrix}
    1 & -1 & -1 \\
    -1 & 1 & -2 \\
    1 & -2 & 1 \\
\end{bmatrix}
$

для $\forall \vec{x}^{(0)}$ сходиться при $\tau = \frac{1}{2}$?

<center><h1> Решение </h1></center>

1) Заметим, что $\lambda = 2$ - собственное число матрицы $\textbf{A}$. Тогда $2 + \lambda_2 + \lambda_3 = 6$, $\det(A) = 2\lambda_2\lambda_3 = 4$ $\implies$ $\lambda_{2, 3} = 2 \pm \sqrt{2} $. То есть имеем
$$0 < \tau = \frac{1}{2} = \frac{2}{4} < \frac{1}{2 + \sqrt{2}}$$
откуда следует, что метод будет сходиться $\forall \vec{x}^{(0)}$.

2) Заметим, что $\lambda_{1} = 1, \lambda_2 = 3$. Тогда $\lambda_3 = 2$. $\forall \lambda :|1 - \tau\lambda| \leq \frac{1}{2} < 1 $, т.е. метод сходится $\forall \vec{x}^{(0)}$.

3) Заметим, что $\lambda = -1$ - собственное число матрицы $\textbf{A}$. Методом подбора находим, что $\lambda = 1$ тоже является собственным числом рассматриваемой матрицы. Тогда $\lambda_3 = 3$. Для $\lambda = -1$ выполняется $|1 - \tau\lambda| = \frac{3}{2} \nless 1$, т.е. метод $\underline{не}$ будет сходиться $\forall \vec{x}^{(0)}$.

___

__Определение (S(c + 0))__: Пусть $c$ – действительное число и $f(c) \neq 0$. Через выражение $S(c + 0)$ обозначим число перемен знаков в последовательности функции и её производных $\{f_x^{(i)}(c)\}_{i = 0}^n$ в точке $c$, где $f_x^{(i)}(c + 0) = 0$ присваивается знак в точке $x$, сколь угодно близкой к $c$ справа, а $f_x^{(i)}(c - 0) = 0$ - слева.

__Определение (ряд Штурма)__: На отрезке действительной оси $[a, b]$ составим последовательность $(f_i)_{i = 0}^n, f_0 = f(x), f_1 = f_x'(x), f_2 = -Res\{f_0/f_1\}, \dots, f_n = -Res\{f_{n - 2}/f_{n - 1}\}$, где $Res\{f_p(x)/f_q(x)\}$ - остаток от деления многочлена $f_p(x)$ на многочлен $f_q(x)$, и назовём её рядом Штурма.

__Теорема (Декарта)__: Число положительных действительных корней уравнения $f(z) = a_0z^n + \dots + a_n = 0$ с учётом их кратности либо равно числу перемен знаков в последовательности $\{a_i\}_{i = 0}^n$, коэффициентов, причём коэффициенты, равные нулю, не учитываются, либо меньше этого числа на чётное число.



__Теорема (Бюдана-Фурье)__: Если $a$ и $b$ ($a < b$) не являются корнями уравнения $f(z) = a_0z^n + \dots + a_n = 0$, то число действительных корней этого уравнения с учётом их кратности на $[a, b]$ равняется разности между числом перемен знаков в последовательностях $S(a + 0)$ и $S(b - 0)$ или меньше этого числа на чётное число
$$\Delta = S(a + 0) - S(b - 0).$$

__Теорема (Штурма)__: Число действительных корней уравнения $f(z) = a_0z^n + \dots + a_n = 0$ на $[a, b]$, если среди них нет кратных, равняется разности между числом перемены знаков в ряде Штурма при $x = a$ и $x = b$.
___

<center> <h1> Задача 3.1.1 </h1> </center>

Определить число действительных корней с использованием: 
<style type="text/css">
    ol { list-style-type: upper-alpha; }
</style>    
    а) теоремы Декарта
    б) теоремы Бюдана–Фурье
    в) теоремы Штурма

и локализовать их для $f(x) = 20x^3 - 4x^2 - 5x + 1 = 0$.

<center> <h1> Решение </h1> </center>

<p>a)</p>
    По теореме Декарта число положительных действительных корней равно корней числу перемен знаков в последовательности {20, -4, -5, 1}, т.е. 2. Их могло вообще не быть (т.е. быть на 2 меньше), но в точке $x = 1/2$ точно находится 1 корень, а значит их 2. В силу основной теоремы алгебры, на отрицательной оси также должен быть 1 корень. Т.е. $f(x)$ имеет 3 действительных корня.

<p>б)</p>
    Несложно показать, что все корни уравнения $f(x) = 0$ находятся на отрезке $[-1, 1]$. Возьмём производные:
    $$f_x' = 60x^2 - 8x - 5 \implies f_x'(-1) > 0 \wedge f_x'(1) > 0$$
    $$f_{xx}'' = 120x - 8 \implies f_{xx}''(-1) < 0 \wedge f_{xx}''(1) > 0$$
    $$f_{xxx}''' = 120 \implies f_{xxx}'''(-1) = f_{xxx}'''(1) > 0$$
    То есть $\Delta = 3 - 0 = 3$. Два корня ($x = 0.5$ и некий корень на $[-1, 0]$) точно есть, тогда и третий корень находится на заданном промежутке.

<p>в)</p>
    Составим ряд Штурма:
    $$f_0 = f(x) = 20x^3 - 4x^2 - 5x + 1 \quad f_1 = f'(x) = 60x^2 - 8x - 5$$
    $$f_2 = \frac{158}{45}x - \frac{8}{9} \quad f_3 = C > 0$$
    Для них имеем:
    $$f_0(-1) < 0 \wedge f_0(1) > 0 \quad f_1(-1) > 0 \wedge f_1(1) > 0$$
    $$f_2(-1) < 0 \wedge f_2(1) > 0 \quad f_3(-1) = f_3(1) > 0$$
    То есть $\Delta = 3 - 0 = 3$ корня на отрезке $[-1, 1]$.

<center> <h1> Задача 3.2.2 </h1> </center>
Локализовать первый положительный корень уравнения $\cos(x) = \frac{1}{x}$ . Предложить метод простой итерации для вычисления этого корня и проверить выполнение условия сходимости в области локализации. 

<center> <h1> Решение </h1> </center>

Зададим функцию $f(x) = \cos x - 1/x$. Графически убедимся, что 1-ый положительный корень лежит в промежутке $[\pi, 2\pi]$. Для метода простой итерации рассмотрим функцию $x = 2\pi - \arccos(1/x)$. Для функции $\phi(x) = 2\pi - \arccos(1/x)$ имеем $|\phi'(x)| = \left|\frac{1}{\sqrt{1 - \frac{1}{x^2}}}\frac{1}{x^2}\right| = \frac{1}{x\sqrt{x^2 - 1}} \leq \frac{1}{\pi^2 - 1} < 1$, т.е. такой итерационный процесс сойдётся.

Теперь предоставим код.

In [20]:
from iter_process import iter_process

x = iter_process(lambda x: 2 * np.pi - np.arccos(1/x), x_0 = 2 * np.pi, iter_num = 20)
print(x)

4.917185925287132


<center> <h1> Задача 3.2.12 </h1> </center>

Уравнение ${\rm e}^x = x^2$ среди своих корней имеет корень $x_{*} \approx -0.7$. Для нахождения корня предлагается использовать один из следующих методов простой итерации: 

1. $x_{n + 1} = 2\ln(-x_n)$
2. $x_{n + 1} = -\sqrt{{\rm e}^{x_n}}$
3. $x_{n + 1} = {\rm e}^{x_n}/x_n$
4. $x_{n + 1} = \frac{1}{6}x_n - \frac{5}{6}{\rm e}^{\frac{x_n}{2}}$.

Оценить скорость сходимости каждого метода и выбрать наилучший.

<center> <h1> Решение </h1> </center>

Рассмотрим функции $f_1(x) = 2\ln(-x)$, $f_2(x) = -\sqrt{{\rm e}^{x}}$, $f_3(x) = {\rm e}^x/x$, $f_4(x) = \frac{1}{6}x - \frac{5}{6}{\rm e}^{\frac{x}{2}}$. Для них имеем
$$|f_1'(x_{*})| = -\frac{2}{x_{*}} > 1$$
$$|f_2'(x_*)| = \frac{1}{2}{\rm e}^{\frac{x_*}{2}} \approx 0.248$$
$$|f_3'(x_*)| = e^{x_*}\left(\frac{1}{x_*^2} - \frac{1}{x_*}\right) > 1$$
$$|f_4'(x_*)| = \frac{5}{12}{\rm e}^{\frac{x}{2}} - \frac{1}{6} \approx 0.127$$

То есть первый и третий методы не сходятся в окрестности корня $x_* \approx 0.7$, а наибольшую скорость сходимости обеспечивает четвёртый метод.

Приведём код, демонстрирующий выкладки.

In [21]:
x_0 = -0.7
iter_num = 5

x_1 = iter_process(lambda x: 2 * np.log(-x), x_0, iter_num)
x_2 = iter_process(lambda x: -np.sqrt(np.exp(x)), x_0, iter_num)
x_3 = iter_process(lambda x: np.exp(x)/x, x_0, iter_num)
x_4 = iter_process(lambda x: x/6 - 5 * np.exp(x/2)/6, x_0, iter_num)
print("iter_num: {} \nf_1: {} \nf_2: {} \nf_3: {} \nf_4: {}".format(iter_num, x_1, x_2, x_3, x_4), end = '\n\n')

iter_num = 20
x_2 = iter_process(lambda x: -np.sqrt(np.exp(x)), x_0, iter_num)
x_4 = iter_process(lambda x: x/6 - 5 * np.exp(x/2)/6, x_0, iter_num)
print("iter_num: {} \nf_2: {} \nf_4: {}".format(iter_num, x_2, x_4), end = '\n\n')

print("true root: -0.70346742249839165")

iter_num: 5 
f_1: -1.4445191744260453 
f_2: -0.7034861016311978 
f_3: -0.7548319969736612 
f_4: -0.7034675347739221

iter_num: 20 
f_2: -0.7034674224954771 
f_4: -0.7034674224983917

true root: -0.70346742249839165


<center> <h1> Задача 3.2.22 </h1> </center>

Отделить действительные корни, выбрать точку начального приближения, написать итерационную формулу метода Ньютона для уточнения одного из действительных корней уравнения, проверить выполнение условий сходимости и привести оценку достижения заданной точности при вычислениях $x^2 - \cos x = 0$.

<center> <h1> Решение </h1> </center>

Графически убеждаемся, что уравнение имеет только 2 корня в промежутках $[-\frac{\pi}{2}, -\frac{\pi}{6}]$ и $[\frac{\pi}{6}, \frac{\pi}{2}]$. Уточним значение положительного корня. Точку начального приближения выберем так, чтобы $f(x_0) \cdot f''(x_0) > 0$. Для этого посчитаем сначала $f'(x), f''(x)$:

$$f'(x) = 2x + \sin x$$
$$f''(x) = 2 + \cos x$$

Можно взять $x = \frac{\pi}{3}$. Тогда $f(\pi/3)f''(\pi/3) = (\frac{\pi^2}{3} - \cos\frac{\pi}{3})(2 + \cos(\frac{\pi}{3})) > 0$. То есть условие сходимости в этом случае выполняется.

Заданная точность будет достигнута при 

$$|x_n - x_*| \leq \frac{M_2}{2m_1}(x_n - x_{n - 1})^2 < \epsilon,$$
где $M_2 = \max\limits_x |f''(x)| = 2 + \frac{\sqrt{3}}{2}$, $m_1 = \min\limits_x |f'(x)| = \frac{\pi}{3} + \frac{1}{2}$. Тогда

$$|x_n - x_{n - 1}| < \sqrt{\frac{2\pi + 6}{12 + 3\sqrt{3}}\epsilon} \approx 0.845\sqrt{\epsilon}.$$

Напишем формулу метода Ньютона для уточнения корня:

$$x_{n + 1} = x_n - \frac{x_n^2 - \cos x_n}{2x_n + \sin x_n}$$

Теперь протестируем формулу.

In [25]:
from iter_process import epsilon_iter_process

x_0 = np.pi/3
sqrt_epsilon = 1e-5 # Точность 10 цифр

x, iter_num = epsilon_iter_process(lambda x: x - (x ** 2 - np.cos(x))/(2 * x + np.sin(x)),
                                   x_0 = x_0,
                                   epsilon = 0.845161799 * sqrt_epsilon)
print(f"number of iterations: {iter_num}")
print(f"stated precision: {np.floor(np.log10(np.abs(sqrt_epsilon ** 2))).astype(int)}")
print(f"x_{iter_num}: {x}")
print("True root: 0.82413231230252242296")

number of iterations: 4
stated precision: -10
x_4: 0.8241323123025232
True root: 0.82413231230252242296


Заметим, что, хоть требуемая точность и 10 цифр, полученный результат даёт точность в 14 цифр.

<center> <h1> Задача 3.2.35 </h1> </center>

Для нахождения $a^{\frac{1}{5}}$ предлагается использовать итерационный процесс $x_{n + 1} = Ax_n + \frac{Ba}{x_n^4} + \frac{Ca^2}{x_n^9}$. Найти значения параметров $A, B, C$, обеспечивающих максимальный порядок сходимости.

<center> <h1> Решение </h1> </center>

Рассмотрим $\phi(x) = Ax + \frac{Ba}{x^4} + \frac{Ca^2}{x^9}$.

$$\frac{\phi(a^{\frac{1}{5}})}{a^{\frac{1}{5}}} = A + B + C = 1$$
$$\phi'(a^{\frac{1}{5}}) = \left(A - \frac{4Ba}{x^5} - \frac{9Ca^2}{x^{10}}\right)|_{x = a^{\frac{1}{5}}} = A - 4B - 9C = 0$$
$$\phi''(a^{\frac{1}{5}}) = \left(\frac{20Ba}{x^6} + \frac{90Ca^2}{x^{11}}\right)|_{x = a^{\frac{1}{5}}} = \frac{20B}{a^{\frac{1}{5}}} + \frac{90C}{a^{\frac{1}{5}}} = 0$$
$$\phi'''(a^{\frac{1}{5}}) = \left(-\frac{120Ba}{x^7} - \frac{990Ca^2}{x^{12}}\right)|_{x = a^{\frac{1}{5}}} = -\frac{120B}{a^{\frac{2}{5}}} -\frac{990C}{a^{\frac{2}{5}}} \neq 0$$

Откуда получим частное решение $A = \frac{18}{25}; B = \frac{9}{25}; C = -\frac{2}{25}$. Максимальный порядок сходимости - 3.

Протестируем полученные результаты.

In [23]:
a = np.longdouble(1.1)
A = 18 / 25
B = 9 / 25
C = -2 / 25

x = iter_process(lambda x: A * x  + B * a / (x ** 4) + C * (a ** 2) / (x ** 9), x_0 = a, iter_num = 10)
print(f"Iter_process result ({iter_num} iterations):", x)
print("True result:", a ** (1/5))

Iter_process result (4 iterations): 1.0192448764914566
True result: 1.0192448764914566


<center> <h1> Задача 3.2.36 </h1> </center>

Определить порядок сходимости итерационного метода для нахождения корня $1 + \sqrt{2}$ уравнения $x^2 - 2x - 1 = 0$:

$$x_{n + 1} = \frac{3}{8}x_n + \frac{5}{8} + \frac{3}{2(x_n - 1)} - \frac{1}{2(x_n - 1)^3}.$$

<center> <h1> Решение </h1> </center>

Рассмотрим $\phi(x) = \frac{3}{8}x + \frac{5}{8} + \frac{3}{2(x - 1)} - \frac{1}{2(x - 1)^3}$, $x_* = 1 + \sqrt{2}$

Имеем:

$$\phi(x_*) = 1 + \frac{3}{8}\sqrt{2} + \frac{3}{2\sqrt{2}} - \frac{1}{4\sqrt{2}} = 1 + \frac{1}{8}\left(3\sqrt{2} + 6\sqrt{2} - \sqrt{2}\right) = 1 + \sqrt{2}$$
$$\phi'(x_*) = \left(\frac{3}{8} - \frac{3}{2(x - 1)^2} + \frac{3}{2(x - 1)^4}\right)_{x = x_*} = 0$$
$$\phi''(x_*) = \left(\frac{3}{(x - 1)^3} - \frac{6}{(x - 1)^5}\right)_{x = x_*} = 0$$
$$\phi'''(x_*) = \left(-\frac{9}{(x - 1)^4} + \frac{30}{(x - 1)^6}\right)_{x = x_*} = -\frac{9}{4} + \frac{30}{8} = \frac{3}{2} \neq 0$$

Т.е. предложенный итерационный метод имеет 3 порядок сходимости.

<center> <h1> Задача 3.2.44 </h1> </center>

Предложить итерационный метод третьего порядка сходимости для нахождения корня $\sqrt[3]{5} - 1$ уравнения:

$$x^3 + 3x^2 + 3x - 4 = 0.$$

<center> <h1> Решение </h1> </center>

Для третьего порядка сходимости мы должны иметь следующее рекуррентное соотношение (при условиях непрерывности $f(x)$, существовании нужного числа производных на отрезке локализации и $f_x'(x) \neq 0$ на нём):

$$x_{n + 1} = x_n - \frac{f_n}{(f_x')_n} - \frac{(f_{xx}'')_n}{2(f_x')_n^3}f_n^2.$$

Для нашей функции $f(x) = x^3 + 3x^2 + 3x - 4 = (x + 1)^3 - 5$ имеем

$$x_{n + 1} = x_n - \frac{(x_n + 1)^3 - 5}{3(x_n + 1)^2} - \frac{6(x_n + 1)}{54(x_n + 1)^6}\left((x_n + 1)^3 - 5\right)^2 = $$
$$ = x_n - \frac{1}{3}(x_n + 1) + \frac{5}{3(x_n + 1)^2} - \frac{1}{9(x_n + 1)^5}\left((x_n + 1)^6 - 10(x_n + 1)^3 + 25\right) = \frac{5}{9}x_n - \frac{4}{9} + \frac{25}{9(x_n + 1)^2} - \frac{25}{9(x_n + 1)^5}$$

Заметим, что в качестве отрезка локализации в данной задаче можно выбрать отрезок $[0, 1]$, в котором и лежит $x_*$. На этом отрезке по теореме Декарта есть только 1 корень, коим и является $x_*$.

<center> <h1> Задача 3.3.2 </h1> </center>

Предложить метод простой итерации и проверить выполнение достаточного условия сходимости последовательных приближений к корню $x^* \in [\pi, 3\pi/2]$; $y^* \in [\pi/4, \pi/2]$ системы нелинейных уравнений:

$$
\begin{cases}
    x = \tan{y} ,\\
    y = \tan{x} .\\
\end{cases}
$$

<center> <h1> Решение </h1> </center>

$\DeclareMathOperator{\arccot}{arccot}$

$$x = \tan y = \cot(\frac{\pi}{2} - y) \iff y = \frac{\pi}{2} - \arccot x$$
$$y = \tan x = \cot(\frac{3\pi}{2} - x) \iff x = \frac{3\pi}{2} - \arccot y$$

То есть итерационный процесс будет иметь вид:

$$
\begin{cases}
    x_{n + 1} = \frac{3\pi}{2} - \arccot y_n \\
    y_{n + 1} = \frac{\pi}{2} - \arccot x_n \\
\end{cases}
$$

Матрица Якоби имеет вид:

$$J = 
\begin{bmatrix}
    0 & \frac{1}{1 + y^2}\\
    \frac{1}{1 + x^2}& 0 \\
\end{bmatrix}
$$

Для неё выполняется
$$||J|| \leq \frac{1}{1 + \pi^2} + \frac{1}{1 + \frac{\pi^2}{16}} < 1$$

То есть достаточное условие сходимости для предложенного итерационного процесса выполняется.

Теперь протестируем наши результаты. В модуле ```numpy``` нет функции $\arccot$, поэтому воспользуемся тем, что в рассматриваемых промежутках верно тождество $\arccot{x} = \arctan{\frac{1}{x}}$.

In [24]:
x_0 = np.array([5 * np.pi / 4, 3 * np.pi / 4])
iter_num = 100

x, y = iter_process(lambda x: (3 * np.pi / 2, np.pi / 2) - np.arctan(1/x[::-1]), x_0 = x_0, iter_num = iter_num)
print(f"Iter_process solution: ({np.float64(x):.15f}, {np.float64(y):.15f})")
print("True solution: (4.067588865765862, 1.329731220678945)")

Iter_process solution: (4.067588865765862, 1.329731220678945)
True solution: (4.067588865765862, 1.329731220678945)


<center> <h1> Задача 3.3.8 </h1> </center>

Написать формулу метода Ньютона для итерационного решения системы нелинейных уравнений 

$$
\begin{cases}
    \sin(x + y) - 1.5x = 0 ,\\
    x^2 + y^2 = 1 .\\
\end{cases}
$$

<center> <h1> Решение </h1> </center>

Матрица Якоби имеет вид:

$$J = 
\begin{bmatrix}
    \cos(x + y) - 1.5 & \cos(x + y) \\
    2x & 2y \\
\end{bmatrix}
$$

Тогда имеем следующую формулу для метода Ньютона:

$$
\begin{bmatrix}
    x_{n + 1} \\
    y_{n + 1} \\
\end{bmatrix} 
=
\begin{bmatrix}
    x_{n} \\
    y_{n} \\
\end{bmatrix}
- \frac{1}{2(y_n - x_n)\cos(x_n + y_n) - 3y_n}
\begin{bmatrix}
    2y_n & -\cos(x_n + y_n) \\
    -2x_n & \cos(x_n + y_n) - 1.5\\
\end{bmatrix}
\begin{bmatrix}
    \sin(x_n + y_n) - 1.5x_n \\
    x_n^2 + y_n^2 - 1 \\
\end{bmatrix}
.$$